proposals/0000-markdown-content-redesign.md

```diff
@@ -0,0 +1,152 @@
+- **Infinite Loops:** By rendering during import, we also introduce a risk for infinite loops when we call `fetchContent()` in a Markdown layout used by a fetched page, bouncing between render and `fetchContent()` infinitely.
+
+
+# Detailed design
+
+## `Astro.glob()`
+
+```diff
+// 1.
+// We convert `Astro.glob()` calls at the point of the call, so that Vite
+// can do its glob-magic without us re-implementing the complexity on our end.
+// This is currently done on Astro.fetchContent(), so no change needed to existing behavior.
+- Astro.glob('./foo/*.md');
++ Astro.glob(import.meta.globEager('./foo/*.md'));
+```
+
+```ts
+// 2.
+Astro.glob = function<T=any>(importMetaGlobResult: Record<string, any>): Promise<T[]> {
+  // Convert the `import.meta.globEager` result into an array.
+  let allEntries = [...Object.values(importMetaGlobResult)];
+  // Report an error if no objects are returned.
+  // TODO: This may no longer be needed, since we changed Vite logging from error -> warn.
+  if (allEntries.length === 0) {
+    throw new Error(`Astro.glob() - no matches found.`);
+  }
+  // NOTE: This API was designed to be async, however we convert its argument to a resolve `globEager` 
+  // object at compile time. We fake asynchrony here so that this API can still become async in the 
+  // future if we ever move off of `import.meta.globEager()`. This should not impact users too much.
+  return Promise.resolve(allEntries);
+}
+```
+
+- This replaces `Astro.fetchContent()` as the new preferred API
+- This should support 99% of usage, especially when importing Markdown.
+- This is optional: for more advanced use-cases we will still document the lower `import.meta.glob` & `import.meta.globEager` Vite APIs as advanced fallbacks.
+- This is Astro-only: Glob imports inside JS and framework components are also considered advanced usage to use the Vite APIs.
+
+## Deferred Import Implementation
+
+- All Markdown files (ex: `src/posts/foo.md`) are resolved and loaded by an internal `'astro:markdown'` plugin.
+- This logic is implemented as v2.0 of that same internal plugin.
+
+1. `src/posts/foo.md?import`
+    1. loaded from file system
+    2. frontmatter is parsed from the file
+    3. a JS module is returned with the following JS:
+    ```js
+    `
+    // Static:
+    export const frontmatter = ${JSON.stringify(frontmatter)};
+    export const file = ${JSON.stringify(id)};
+    export const url = ${JSON.stringify(url || undefined)};
+
+    // Deferred:
+    export default async function load(...args) {
+        return (await import(${JSON.stringify(fileId + '?content')}));
+    };
+    export function getContent() {
+        return load().then((m: any) => m.default)
+    }
```

Comment:
I think it would be preferred if the API just exposed the component itself rather than needing to call a getter and render it in an expression.

```suggestion
    export function Content(...args: any) {
        return load().then((m: any) => m.default(...args))
    }
    Content.isAstroComponentFactory = true;
```

## Usage

**Before**
```astro
---
const firstPost = await import('../posts/first-post.md');
---
<!-- Data is parsed from frontmatter without requiring a render -->
<h1>{firstPost.frontmatter.title}</h1>
<p>Author: {firstPost.frontmatter.author}</p>
<p><a href={firstPost.url}>Permalink</a></p>
<!-- Defer rendering as late as possible for performance -->
<article>{firstPost.getContent()}</article>
```

**After**
```astro
---
import { Content, frontmatter, file, url } from "../posts/first-post.md";
---
<!-- Data is parsed from frontmatter without requiring a render -->
<h1>{frontmatter.title}</h1>
<p>Author: {frontmatter.author}</p>
<p><a href={url}>Permalink</a></p>
<!-- Still defers rendering as late as possible for performance, but the user doesn't need to worry about that -->
<article><Content /></article>
```

Reply:
Basically, I think we can delay the rendering for performance reasons without passing that cost onto the user. Astro will `await` the component before attempting to render it, which amounts to automatic unwrapping.

A component-based API makes more sense because `firstPost.getContent()` wouldn't be inspectable like users might think.

Reply:
Ah yea, I'm game for this 👍 That feels much more natural

Reply:
Well, actually with static ESM imports that feels more natural, but in all other cases I'm not sure it does. Look what it does to the example we already have:

```astro
---
const firstPost = await import('../posts/first-post.md');
---
<firstPost.Content />
```

or, for the collection/glob use-case:

```astro
---
const posts = await Astro.glob('../posts/first-post.md');
---
{posts.map(post => <post.Content />)}
```

We do support this syntax, but we don't push users to use it very often. Are we okay with that? 

Also, does `<Content />` give us the ability to pass props to the layout that the markdown page uses? That could be a deciding factor here, if we can think of a use-case there that we want to support.

Reply:
I think this is fine, since you can destructure `{ Content }` quite easily in the `.map` example. I was also wondering about props as well. I don't think this would be possible, since this would change frontmatter from a compile-only step to a render step 😬

Reply:
Accepted suggestion into RFC!